src/views/honeysql/util.clj

```diff
@@ -32,31 +32,35 @@
        (keep isolate-tables)
        (swap! tables #(into %1 %2))))
 
-(defn insert-tables
-  [query]
-  (some->> query :insert-into first-leaf vector))
+(defn from-join-tables
+  [node from tables]
+  (from-tables from tables)
+  (doseq [join-key [:join :left-join :right-join :full-join]]
+    (when-let [join (get node join-key)] (join-tables join tables))))
 
-(defn update-tables
-  [query]
-  (if-let [v (:update query)] [v] []))
+(defn insert-tables
+  [insert tables]
+  (when-let [v (first-leaf insert)]
+    (swap! tables conj v)))
 
 (defn delete-tables
-  [query]
-  (if-let [v (:delete-from query)] [v] []))
+  [node delete-from tables]
+  (from-tables [delete-from] tables)
+  (when-let [v (:using node)]
+    (from-tables v tables)))
 
 (defn query-tables
   "Return all the tables in an sql statement."
   [query]
   (let [tables (atom #{})
         get-tables (fn [node]
                      (when (map? node)
-                       (when-let [from (:from node)]
-                         (from-tables from tables)
-                         (doseq [join-key [:join :left-join :right-join :full-join]]
-                           (when-let [join (get node join-key)] (join-tables join tables)))))
+                       (let [{:keys [from insert-into update delete-from]} node]
+                         (when update      (from-join-tables node [update] tables))
+                         (when from        (from-join-tables node from tables))
+                         (when insert-into (insert-tables insert-into tables))
+                         (when delete-from (delete-tables node delete-from tables))))
                      node)]
     (clojure.walk/postwalk get-tables query)
-    (-> @tables
-        (into (insert-tables query))
-        (into (update-tables query))
-        (into (delete-tables query)))))
+    @tables))
+
```

Comment:
I did not consider this prior, however I wonder whether this can be implemented in a functional manner - ie. without an atom. To claim that this stateful derivation is not "idiomatic" is a cop-out - but the same result can be achieved in a pure-functional manner.

This would require a mild reformat of the helper functions above to chain tables through the calls, but it can be done:
```
(defn delete-tables
  [node delete-from tables]
  (cond->> (from-tables [delete-from] tables)
    (:using node)
    (from-tables (:using node))))
```
... and so on ...

As it stands, because you are passing an atom through these functions, they are inherently stateful. that is to say - they cannot be reused without:
- knowing the expected shape of the internals of the atom
- having at atom to get output out of the function.

As it stands, all of the functions contained herein should be made private BECAUSE they are stateful. Thus, this may not be the right approach.

You may do well to reconsider this approach in favour to pure functions.

Reply:
fixed https://github.com/kirasystems/views-honeysql/pull/12/commits/a9ab9196672969c6a05f238c8f460d36bf61c88b